### PR TITLE
DEV2-2537: IJ plugin should install the binaries from the self-hosted updsate service

### PR DESCRIPTION
**Pull request by assapir**
Tested locally: 
Was able to get the version, download and run TabNine
```
00:00:00 /home/assaf/.tabnine/4.4.277/x86_64-unknown-linux-musl/TabNine --client IntelliJIDEA --no-lsp true --cloud2_url=http://localhost:3000 --client-metadata pluginVersion=1.0.0 clientIsUltimate=false clientChannel=onprem pluginUserId=ca109c96-95ad-49dc-ae13-a36185156caa debounceValue=300 clientVersion=2019.3 clientApiVersion=IC-193.5233.102 ide-restart-counter=0

```

### CODE REVIEW

File: src/main/java/com/tabnine/binary/fetch/BundleDownloader.java [hunk 1]
```diff
@@ -23,6 +26,15 @@ public BundleDownloader(TempBundleValidator validator, GeneralDownloader downloa
 
   public Optional<BinaryVersion> downloadAndExtractBundle(String version) {
     String urlString = String.join("/", getBundleServerUrl(), version, TARGET_NAME, "TabNine.zip");
+
+    if (Config.IS_ON_PREM) {
+      if (!StaticConfig.getTabnineEnterpriseHost().isPresent()) {
```
Comment (dimacodota) [inline src/main/java/com/tabnine/binary/fetch/BundleDownloader.java:31]: you can chain `ifPresen(host ->  return host). orElseget( () -> {
Logger.getInstance(getClass()).warn("Not downloading bundle - on prem version but no host is setup!");
        return Optional.empty();}) `

Comment (assapir) [reply src/main/java/com/tabnine/binary/fetch/BundleDownloader.java:31]: I will move it anyway now

File: src/main/java/com/tabnine/general/StaticConfig.java [hunk 2]
```diff
@@ -127,9 +137,9 @@ public static String getTabNineVersionUrl() {
   }
 
   @NotNull
-  public static String getTabNineBundleVersionUrl() {
+  public static String getTabNineBundleVersionUrl() throws NoValidBinaryToRunException {
     return Optional.ofNullable(System.getProperty(REMOTE_VERSION_URL_PROPERTY))
-        .orElse(getBundleServerUrl() + "/version");
+        .orElse(getBundleServerUrl().orElseThrow(NoValidBinaryToRunException::new) + "/version");
```
Comment (yonip23) [inline src/main/java/com/tabnine/general/StaticConfig.java:142]: im not sure about the error handling here - what happens if `NoValidBinaryToRunException` is thrown here? does the user see the ugly popup with this exception?
i think you should instead `warn` log this case, and return `Optional<String>` here, and the user should see the gray (inactive) status bar if the URL is not set (this is already implemented)

File: src/main/java/com/tabnine/general/StaticConfig.java [hunk 3]
```diff
@@ -115,9 +115,19 @@ public static String getServerUrl() {
         .orElse("https://update.tabnine.com");
   }
 
-  public static String getBundleServerUrl() {
-    return Optional.ofNullable(System.getProperty(REMOTE_BASE_URL_PROPERTY))
-        .orElse("https://update.tabnine.com/bundles");
+  public static Optional<String> getBundleServerUrl() {
+    if (Config.IS_ON_PREM) {
+      if (!StaticConfig.getTabnineEnterpriseHost().isPresent()) {
```
Comment (yonip23) [inline src/main/java/com/tabnine/general/StaticConfig.java:120]: extract `StaticConfig.getTabnineEnterpriseHost()` to variable?

**Review by yonip23 — COMMENTED** [review]
not sure about this (see comment)

File: src/test/java/com/tabnine/binary/BinaryVersionFetcherTests.java [hunk 4]
```diff
@@ -30,10 +30,10 @@ public class BinaryVersionFetcherTests {
   @InjectMocks private BinaryVersionFetcher binaryVersionFetcher;
 
   @BeforeEach
-  public void setUp() throws BackingStoreException {
+  public void setUp() throws BackingStoreException, NoValidBinaryToRunException {
     Preferences preferences = Preferences.userNodeForPackage(BootstrapperSupport.class);
     preferences.clear();
-    when(binaryRemoteSource.fetchPreferredVersion(StaticConfig.getTabNineBundleVersionUrl()))
+    when(binaryRemoteSource.fetchPreferredVersion(anyString()))
```
Comment (yonip23) [inline src/test/java/com/tabnine/binary/BinaryVersionFetcherTests.java:36]: its better if you have the specific string to compare against - why did you change this? what broke the test?
same question for all the `anyString` changes

**Review by yonip23 — COMMENTED** [review]
see comments
also, please provide some evidence of manual testing (either here / slack / doc / whatever)

File: src/test/java/com/tabnine/binary/BinaryVersionFetcherTests.java (hunk 4 above)
Comment (assapir) [reply src/test/java/com/tabnine/binary/BinaryVersionFetcherTests.java:36]: because now I need to mock it inside

**Review by yonip23 — APPROVED** [review]
lgtm, well done for winning this battle with java 💪🏻

**Comment by assapir** [comment]
> lgtm, well done for winning this battle with java 💪🏻

Thanks for the help ❤️

File: src/main/java/com/tabnine/general/StaticConfig.java [hunk 5]
```diff
@@ -115,21 +114,27 @@ public static String getServerUrl() {
...
+      }
+      return Optional.of(String.join("/", tabnineEnterpriseHost.get(), "update", "bundles"));
+    }
 
-  @NotNull
-  public static String getTabNineVersionUrl() {
-    return Optional.ofNullable(System.getProperty(REMOTE_VERSION_URL_PROPERTY))
-        .orElse(getServerUrl() + "/version");
+    return Optional.of(
+        Optional.ofNullable(System.getProperty(REMOTE_BASE_URL_PROPERTY))
```
Comment (yanir-codota) [inline src/main/java/com/tabnine/general/StaticConfig.java:128]: what is this property? is it legit for non-onprem?
